decompiler/frontend/binaryninja/handlers/globals.py

```diff
@@ -47,133 +84,207 @@ def __init__(self, lifter):
             FloatType: self._lift_basic_type,
             BoolType: self._lift_basic_type,
             VoidType: self._lift_void_type,
-            ArrayType: self._lift_constant_type,
+            ArrayType: self._lift_array_type,
             PointerType: self._lift_pointer_type,
-            NamedTypeReferenceType: self._lift_named_type_ref,  # Lift DataVariable with type NamedTypeRef
+            NamedTypeReferenceType: self._lift_named_type_ref,
         }
+        self._lifted_globals: dict[int, GlobalVariable] = {}
+        self._view: Optional[BinaryView] = None
+        self._callers: list[int] = []
 
     def register(self):
         """Register the handler at its parent lifter."""
         self._lifter.HANDLERS.update({DataVariable: self.lift_global_variable})
 
+    def _get_gvar_name(self, bninjaName: Optional[str], addr: int) -> str:
+        lifted_names = [v.name for v in self._lifted_globals.values()]
+        if bninjaName is None:
+            return GLOBAL_VARIABLE_PREFIX + f"{addr:x}"
+        if bninjaName in lifted_names:
+            return bninjaName + "_" + f"{addr:x}"  # @Reviewer: Other ideas for distinct name?
+        return bninjaName
+
+    def _build_global_variable(self, name: Optional[str], type: Type, addr: int, init_value, ssa_label: Optional[int]) -> GlobalVariable:
+        vname = self._get_gvar_name(name, addr)
+
+        match init_value:
+            case Expression():
+                vinit_value = init_value
+            case int() | float() | bytes():
+                vinit_value = Constant(value=init_value, vartype=type)
+            case _:
+                raise TypeError(f"Type violation: '{init_value}'")
+
+        self._lifted_globals[addr] = GlobalVariable(
+            name=vname, vartype=type, initial_value=vinit_value, ssa_label=ssa_label, is_constant=self.addr_in_ro_section(self._view, addr)
+        )
+        return self._lifted_globals[addr]
+
     def lift_global_variable(
-        self, variable: DataVariable, view: BinaryView, parent: Optional[MediumLevelILInstruction] = None, caller_addr: int = None, **kwargs
-    ) -> Union[Symbol, UnaryOperation, GlobalVariable, StringSymbol]:
-        """Lift global variables via datavariable type. Check bninja error case + recursive datavariable first"""
-        if not self._addr_in_section(view, variable.address):
+        self, variable: DataVariable, view: BinaryView, parent: Optional[MediumLevelILInstruction] = None, callers: list[int] = [], **kwargs
+    ) -> Union[Constant, Symbol, GlobalVariable]:
+        """Lift global variables via datavariable type"""
+        # Save view for all internal used functions
+        if not self._view:
+            self._view = view
```

Comment:
Again more or less the same reason as above.
Almost all handlers need the view to work and it's kinda annoying to move it around the whole time with `parent` and the `datavariable`.

Furthermore the view will always be the same for a decompiled binary.
I know it disrespect the stateless approach of the lifter, but yeah.... (the state of view does not change :) )

Reply:
Because this is more or less harmless I will accept it if you add a comment near the constructor (stating that the `_view` is set once in `lift_global_variable`).

Reply:
Done.